OpenRA.Game/Input/NamedHotkey.cs

```diff
@@ -0,0 +1,49 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+
+namespace OpenRA
+{
+	/// <summary>
+	/// A reference to either a named hotkey (defined in the game settings) or a statically assigned hotkey
+	/// </summary>
+	public class NamedHotkey
+	{
+		static readonly Func<Hotkey> NoHotkey = () => Hotkey.Invalid;
+
+		readonly Func<Hotkey> getValue;
+
+		public NamedHotkey()
+		{
+			getValue = NoHotkey;
```

Comment:
All `ButtonWidget`s that don't have an explicit key defined will hit this path.  We want them to share a common static notkey (not-a-hotkey) to avoid allocating a bunch of duplicate anonymous functions.